pkg/volume/git_repo/git_repo_test.go

```diff
@@ -96,14 +96,6 @@ func TestPlugin(t *testing.T) {
 					cmd: []string{"git", "clone", "--", gitURL, "target_dir"},
 					dir: "",
 				},
-				{
-					cmd: []string{"git", "checkout", revision},
```

Comment:
why are these test cases removed?

Reply:
Inconsistency between results and expectations due to wrong `expecteds`
Take the case `target-dir` as an example

In the wrong case, the last element in the expecteds slice modifies the previous element
```
		if expected.cmd[1] == "clone" {
			fakeOutputs = append(fakeOutputs, func() ([]byte, []byte, error) {
				// git clone, it creates new dir/files
				os.MkdirAll(filepath.Join(fcmd.Dirs[0], expected.dir), 0750)
				return []byte{}, nil, nil
			})
		} 
``` 
What is actually being passed in is
```
{
					cmd: []string{"git", "reset", "--hard"}.
					dir: "/target_dir".
}.
```
It will create the directory as expected by the error and pass it with no problem
```
(b *gitRepoVolumeMounter) SetUpAt()
```
Finally, it returns the value that matches the wrong expectation

If we add
```
expected := expected
```
what is actually being passed in is
```
{
					cmd: []string{"git", "clone", "--", gitURL, "target_dir"}.
					dir: "".
}.
```
It will just create `fcmd.Dirs[0]`, and the default case of switch in `(b *gitRepoVolumeMounter) SetUpAt()` returns
This time only the first element in `expecteds` matches the test result

Reply:
I was misunderstanding the test, I thought for the given git repo volume source:

```go
v1.GitRepoVolumeSource{
    Repository: gitURL,
    Revision:   revision,
    Directory:  "target_dir",
}
```

The command executed must be: cloning the repo, checkout the specifc revision and then reset the HEAD

Reply:
dropping the test cases is incorrect... the expectations were correct, the fake git driver implemented `clone` incorrectly, and the git plugin was erroring out early.

capturing the error returned by the mount call reveals this:
```diff
 	g := mounter.(*gitRepoVolumeMounter)
 	g.exec = fake
 
-	g.SetUp(volume.MounterArgs{})
+	err := g.SetUp(volume.MounterArgs{})
+	if err != nil {
+		allErrs = append(allErrs, err)
+	}
 
 	if fake.CommandCalls != len(expecteds) {
 		allErrs = append(allErrs,
```

and the fix is to make the fake clone implementation create the expected subdirectory:
```diff
 	var fakeOutputs []fakeexec.FakeAction
 	var fcmd fakeexec.FakeCmd
 	for _, expected := range expecteds {
+		expected := expected
 		if expected.cmd[1] == "clone" {
+			// Calculate the subdirectory clone would create (if any)
+			// git clone -- https://github.com/kubernetes/kubernetes.git target_dir --> target_dir
+			// git clone -- https://github.com/kubernetes/kubernetes.git            --> kubernetes
+			// git clone -- https://github.com/kubernetes/kubernetes.git .          --> .
+			// git clone -- https://github.com/kubernetes/kubernetes.git ./.        --> .
+			cloneSubdir := filepath.Base(expected.cmd[len(expected.cmd)-1])
+			if cloneSubdir == "kubernetes.git" {
+				cloneSubdir = "kubernetes"
+			}
+
 			fakeOutputs = append(fakeOutputs, func() ([]byte, []byte, error) {
 				// git clone, it creates new dir/files
-				os.MkdirAll(filepath.Join(fcmd.Dirs[0], expected.dir), 0750)
+				os.MkdirAll(filepath.Join(fcmd.Dirs[0], expected.dir, cloneSubdir), 0750)
 				return []byte{}, nil, nil
 			})
 		} else {
```

Reply:
> dropping the test cases is incorrect... the expectations were correct, the fake git driver implemented `clone` incorrectly, and the git plugin was erroring out early.
> 
> capturing the error returned by the mount call reveals this:
> 
> ```diff
>  	g := mounter.(*gitRepoVolumeMounter)
>  	g.exec = fake
>  
> -	g.SetUp(volume.MounterArgs{})
> +	err := g.SetUp(volume.MounterArgs{})
> +	if err != nil {
> +		allErrs = append(allErrs, err)
> +	}
>  
>  	if fake.CommandCalls != len(expecteds) {
>  		allErrs = append(allErrs,
> ```
> 
> and the fix is to make the fake clone implementation create the expected subdirectory:
> 
> ```diff
>  	var fakeOutputs []fakeexec.FakeAction
>  	var fcmd fakeexec.FakeCmd
>  	for _, expected := range expecteds {
> +		expected := expected
>  		if expected.cmd[1] == "clone" {
> +			// Calculate the subdirectory clone would create (if any)
> +			// git clone -- https://github.com/kubernetes/kubernetes.git target_dir --> target_dir
> +			// git clone -- https://github.com/kubernetes/kubernetes.git            --> kubernetes
> +			// git clone -- https://github.com/kubernetes/kubernetes.git .          --> .
> +			// git clone -- https://github.com/kubernetes/kubernetes.git ./.        --> .
> +			cloneSubdir := path.Base(expected.cmd[len(expected.cmd)-1])
> +			if cloneSubdir == "kubernetes.git" {
> +				cloneSubdir = "kubernetes"
> +			}
> +
>  			fakeOutputs = append(fakeOutputs, func() ([]byte, []byte, error) {
>  				// git clone, it creates new dir/files
> -				os.MkdirAll(filepath.Join(fcmd.Dirs[0], expected.dir), 0750)
> +				os.MkdirAll(filepath.Join(fcmd.Dirs[0], expected.dir, cloneSubdir), 0750)
>  				return []byte{}, nil, nil
>  			})
>  		} else {
> ```

Thank you for your help, I have corrected according to the advice